common/internal_model/src/main/scala/uk/ac/wellcome/models/work/internal/Image.scala

```diff
@@ -2,70 +2,140 @@ package uk.ac.wellcome.models.work.internal
 
 import java.time.Instant
 
-sealed trait BaseImage[+State <: DataState] extends HasId[State#Id] {
-  val id: State#Id
-  val locations: List[DigitalLocationDeprecated]
-}
-
-case class UnmergedImage[State <: DataState](
-  id: State#Id,
+case class ImageData[+State](
+  id: State,
   version: Int,
   locations: List[DigitalLocationDeprecated]
-) extends BaseImage[State] {
-  def mergeWith(canonicalWork: SourceWork[State],
-                redirectedWork: Option[SourceWork[State]] = None,
-                modifiedTime: Instant): MergedImage[State] =
-    MergedImage[State](
-      id = id,
-      version = version,
-      modifiedTime = modifiedTime,
-      locations = locations,
-      source = SourceWorks[State](canonicalWork, redirectedWork)
-    )
+) extends HasId[State]
 
-  def mergeWith(source: ImageSource[State],
-                modifiedTime: Instant): MergedImage[State] =
-    MergedImage[State](
-      id = id,
+case class Image[State <: ImageState](
+  version: Int,
+  state: State,
+  locations: List[DigitalLocationDeprecated],
+  source: ImageSource[State#SourceDataState],
+  modifiedTime: Instant
+) {
+  def id: String = state.id
+  def sourceIdentifier: SourceIdentifier = state.sourceIdentifier
+
+  def transition[OutState <: ImageState](args: OutState#TransitionArgs = ())(
+    implicit transition: ImageFsm.Transition[State, OutState],
+    // The transition helper method does not allow transitions across source
+    // DataState boundaries
+    sourceEqualityWitness: ImageSource[State#SourceDataState] =:= ImageSource[
+      OutState#SourceDataState]
```

Comment:
Eh what's this syntax, guess I have to read up on it

Reply:
Is this saying that the `transition` method is only available iif `ImageSource[State#SourceDataState]` is equal to `ImageSource[OutState#SourceDataState]`.

So this is checking that source work is in the same state... could you give an example of where this might not be the case?

Reply:
This is very cool imo, as it's technically not new syntax - it's infix syntax for a typeclass that looks like `Eq[A, B]` and for which the language provides an implicit for any `Eq[A, A]`. I liked [this](https://stackoverflow.com/a/22717040) explanation, or [here](https://www.scala-lang.org/api/current/scala/$eq$colon$eq.html) are the docs, or additionally I found the cats documentation on [their more powerful alternative](https://github.com/typelevel/cats/blob/master/core/src/main/scala/cats/evidence/Is.scala) quite interesting too

Reply:
That explanation is great, I found [this one](https://twitter.github.io/scala_school/advanced-types.html#otherbounds) less verbose and quite useful too.

Just checking my understanding, but it's constraining the source work's `DataType` to ensure that they are the same.

Is this so that you can transition `ImageState.Identified[WorkState.Identified] => ImageState.Augmented[WorkState.Identified]` but not `ImageState.Identified[WorkState.Identified] => ImageState.Augmented[WorkState.Unidentified]`?

Reply:
Exactly - it removes the need for an identity method for `source` on the `Transition` instances like we do [here for works](https://github.com/wellcomecollection/catalogue/blob/master/common/internal_model/src/main/scala/uk/ac/wellcome/models/work/internal/Work.scala#L217).

I didn't change it on works because this PR was big enough already, but I believe that this approach provides the exact same conversion via the same mechanism (ie implicit typeclass resolution), just (a) using a language builtin and (b) using implicit conversion of the `source` so it's invisible.

Reply:
Sorry @jamesgorrie missed your question in your first comment - the only time where the `DataState` would change would be from `Unidentified -> Identified` inside the `ImageState` transition `Initial -> Identified`, and we don't and won't provide a `Transition` for this because it happens automagically inside the ID minter (it's opaque to the services either side).